src/Nethermind/Nethermind.TxPool/TxPoolSender.cs

```diff
@@ -1,28 +1,56 @@
 // SPDX-FileCopyrightText: 2022 Demerzel Solutions Limited
-// SPDX-License-Identifier: LGPL-3.0-only 
+// SPDX-License-Identifier: LGPL-3.0-only
 
 using System;
 using System.Threading.Tasks;
 using Nethermind.Core;
 using Nethermind.Core.Crypto;
+using Nethermind.Crypto;
 
 namespace Nethermind.TxPool
 {
     public class TxPoolSender : ITxSender
     {
         private readonly ITxPool _txPool;
         private readonly ITxSealer _sealer;
+        private readonly INonceManager _nonceManager;
+        private readonly IEthereumEcdsa _ecdsa;
 
-        public TxPoolSender(ITxPool txPool, ITxSealer sealer)
+        public TxPoolSender(ITxPool txPool, ITxSealer sealer, INonceManager nonceManager, IEthereumEcdsa ecdsa)
         {
             _txPool = txPool ?? throw new ArgumentNullException(nameof(txPool));
             _sealer = sealer ?? throw new ArgumentNullException(nameof(sealer));
+            _nonceManager = nonceManager ?? throw new ArgumentNullException(nameof(nonceManager));
+            _ecdsa = ecdsa ?? throw new ArgumentException(nameof(ecdsa));
         }
 
         public ValueTask<(Keccak, AcceptTxResult?)> SendTransaction(Transaction tx, TxHandlingOptions txHandlingOptions)
         {
+            bool manageNonce = (txHandlingOptions & TxHandlingOptions.ManagedNonce) == TxHandlingOptions.ManagedNonce;
+            tx.SenderAddress ??= _ecdsa.RecoverAddress(tx);
+            if (tx.SenderAddress is null)
+                throw new ArgumentNullException(nameof(tx.SenderAddress));
+            if (manageNonce)
+            {
+                tx.Nonce = _nonceManager.ReserveNonce(tx.SenderAddress);
+                txHandlingOptions |= TxHandlingOptions.AllowReplacingSignature;
+            }
+            else
+            {
+                _nonceManager.TxWithNonceReceived(tx.SenderAddress, tx.Nonce);
```

Comment:
Ok I see, now we need `SenderAddress` here too... WDYT @smartprogrammer93 @MarekM25 is it acceptable to always recover address?

Reply:
I think we have to do it anyways